internal/controller/maintenancemode_controller.go

```diff
@@ -0,0 +1,184 @@
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate),
+		).
+		Watches(
+			&v1alpha1.StorageClient{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate, storageClientStatusPredicate),
+		).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes,verbs=get;list;update;create;watch;delete
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclients;storageclaims,verbs=get;list;watch
+
+func (r *MaintenanceModeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx, "MaintenanceMode", req)
+	r.log.Info("Reconciling MaintenanceMode")
+
+	nameToStorageClient := map[string]*v1alpha1.StorageClient{}
+
+	maintenanceModes := &ramenv1alpha1.MaintenanceModeList{}
+	if err := r.list(maintenanceModes); err != nil {
+		r.log.Error(err, "failed to list the Maintenance Modes")
+		return reconcile.Result{}, err
+	}
+
+	for i := range maintenanceModes.Items {
+		mm := &maintenanceModes.Items[i]
+		sc := &v1alpha1.StorageClaim{}
+		// MMode's TargetID is replicationID, which in our case is storageClaim name
+		sc.Name = mm.Spec.TargetID
+		if err := r.get(sc); err != nil {
+			return ctrl.Result{}, err
+		}
+		if sc.Spec.StorageClient == "" {
+			return ctrl.Result{}, fmt.Errorf("StorageClaim %s does not have a StorageClient defined", mm.Name)
+		}
+		clientName := sc.Spec.StorageClient
+		if nameToStorageClient[clientName] == nil {
+			storageClient := &v1alpha1.StorageClient{}
+			storageClient.Name = clientName
+			if err := r.get(storageClient); err != nil {
+				return ctrl.Result{}, err
+			}
+			nameToStorageClient[clientName] = storageClient
+		}
+		if nameToStorageClient[clientName].Status.InMaintenanceMode {
+			if err := r.updateStatusCompletedForMM(mm); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update status for MaintenanceMode %s: %w", mm.Name, err)
+			}
+		}
+	}
+
+	storageClients := &v1alpha1.StorageClientList{}
+	if err := r.list(storageClients); err != nil {
+		r.log.Error(err, "failed to list the Storage Clients")
+		return reconcile.Result{}, err
+	}
+
+	for i := range storageClients.Items {
+		storageClient := &storageClients.Items[i]
+		_, needsMaintenanceMode := nameToStorageClient[storageClient.Name]
+		if needsMaintenanceMode != storageClient.Status.InMaintenanceMode {
+			if err := r.toggleMaintenanceModeForClient(storageClient, needsMaintenanceMode); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MaintenanceModeReconciler) toggleMaintenanceModeForClient(storageClient *v1alpha1.StorageClient, enable bool) error {
+	providerClient, err := providerclient.NewProviderClient(
+		r.ctx,
+		storageClient.Spec.StorageProviderEndpoint,
+		10*time.Second,
```

Comment:
why 10? how did we come up with this number?

Reply:
This was used since start in provider-client communication

Reply:
I am ok with the number, but don't we need a const to make sure all places use the same value?